### ISSUE TITLE
offstrategy compaction races with view building on staging sstables

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#11882]
Offstrategy compaction is triggered after repair and the latter places sstables in the staging subdirectory if view building is required.
If offstrategy compaction is triggered before view building completes, it will pick up the staging sstables if they are in the maintenance sstable set and compact them as any other sstable, placing the output sstables in the main sstables directory.

Although the staging sstables are just marked for deletion and won't be deleted while in use by the view builder, placing the output sstables in the main dir can confuse the view builder since it won't detect the changes in the staging dir vs. the state of the data in the main dir, as computed querying data while excluding the staging sstables.
See https://github.com/scylladb/scylladb/blob/20110bdab406111822e9debb10d953bcfc88755f/db/view/view_update_generator.cc#L161

This is exactly the reason why we don't run cleanup on staging sstables and the reason why https://github.com/scylladb/scylladb/pull/10657 was reverted.
See https://github.com/scylladb/scylladb/pull/10890

**Comment by bhalevy** [comment]
Cc @raphaelsc

**Comment by raphaelsc** [comment]
you're right. we need to postpone offstrategy to after view building completes.

**Comment by raphaelsc** [comment]
Prior to off-strategy, new staging files *had to wait for view building completion* before they could be selected for regular compaction.
Similarly, off-strategy can also wait for view building completion.

We have two alternatives:

Off-strategy before view building:
    - output placed in staging dir
    - output files registered for view building
    - view building completion responsible for moving output away from staging dir

Off-strategy after view building:
    - filter out sstables in staging from off-strategy
    - trigger off-strategy on view building completion
    - input files will be in main directory already, as view building already completed

**Comment by bhalevy** [comment]
I see view building as part of the overall repair process so I think it should be done first.

I think it doesn't benefit from offstrategy compaction either since the staging sstables are disjoint from each other.
Is that right, @nyh?

So view building can defer offstrategy until it's done so offstrategy will start in the later event of 5 minutes after last repair finished or view building finished.  I presume that in the normal case, view building will finish before the 5 minutes delay expire anyhow.
